STYLE_GUIDE.md

```diff
@@ -0,0 +1,88 @@
+
+- Added to the `Imports` or `Suggests` field `DESCRIPTION` file in alphabetical order.
+- In the PR that adds the dependency this should be clearly stated in the PR description along with a justification for the dependency, the number and type of downstream dependencies, and an assessment of the risk of the dependency breaking.
+
+More generally when adding functions from external packages (i.e. even if they are already a dependency) the following should be followed:
+
+- Documented in function documentation using the `@importFrom` tag.
+- Used within functions using the `package::function` format.
+
+## Input types and checking
+
+- We support inputs that are coercible to `data.table` objects using `data.table::as.data.table()`. This includes `data.frame` and `tibble` objects. This should be clearly documentede in the function documentation.
+- Any required inputs should be clearly documented in the function documentation.
+- We use an internal function `coerce_dt()` to check inputs are coercible to `data.table` objects and have the correct columns. This function is used in all functions that take data as input. The following function demonstrates this pattern:
+
+```r
+print_dt <- function(dt) {
+    dt <- epinowcast:::coerce_dt(dt, required_cols = c("date", "cases"))
+    return(dt)
+}
+
+print_dt(mtcars)
+
+print_dt(data.frame(cases = 1, date = Sys.Date()))
+# Error in epinowcast:::coerce_dt(dt, required_cols = c("date", "cases")) : 
+#   The following columns are required: date, cases but are not present among mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb
+# (all `required_cols`: date, cases)
+```
+
+In general, we aim to check the inputs for all external facing functions. This is to ensure that the user is aware of any issues with the input data and to provide a consistent error message. See the documentation for `coerce_dt()` for more details. It may also be helpful to review usage in the package more widely, for this `data-converters.R` is a sensible place to start.
+
+- For external facing functions `coerce_dt()` should generally not update by reference (i.e. `copy = TRUE` should be set, the default). In cases where users may benefit from updating by reference the external function should pass through the `copy` argument to `coerce_dt()`.
+- For internal functions `coerce_dt()` should generally update by reference (i.e. `copy = FALSE` should be set) when used internally. This is to avoid unnecessary copying of data.
+
+## Internal data manipulation
+
+- `data.table` objects are used for internal data manipulation. If you are unfamiliar with `data.table` please see the [documentation](https://rdatatable.gitlab.io/data.table/index.html) and [cheatsheet](https://s3.amazonaws.com/assets.datacamp.com/img/blog/data+table+cheat+sheet.pdf). Prototype code may be written with other tools but will generally need be refactored to use `data.table` before submission (in PRs where help is needed with this please clearly state this).
```

Comment:
I think there's also a worthwhile bullet here on *how* we write data.table.

I'm very inclined to bracket chaining approaches (performance-oriented). You seem to use one-liners with re-assignment (maintainability-oriented).

As much as I do not look forward to losing on this issue, I think you should pick one.

Reply:
I shouldn't do anything. We can make a decision but I think that might need to be pushed to a new issue so others can weigh in. I agree on the relative trade off. Really I am waiting for the DT sugar to be released and then we can move to that with the base R pipe to get to some middle ground.

Reply:
mmm, the preferred use of the pipe and lambda's in the code should definitely be a style guide item as well.

Reply:
we don't want them used yet though as the new version of `data.table` hasn't been realised. Agree will need an issue to discuss when it is.